drivers/shared/executor/executor_test.go

```diff
@@ -336,3 +339,81 @@ func setupRootfsBinary(t *testing.T, rootfs, path string) {
 	err = os.Link(src, dst)
 	require.NoError(t, err)
 }
+
+// TestExecutor_Cleanup_Running asserts that CleanupExecutor can kill
+// a running task
+func TestExecutor_Cleanup_Running(pt *testing.T) {
+	pt.Parallel()
+	for name, factory := range executorFactories {
+		pt.Run(name, func(t *testing.T) {
+			require := require.New(t)
+			execCmd, allocDir := testExecutorCommand(t)
+			execCmd.Cmd = "/bin/sleep"
+			execCmd.Args = []string{"100"}
+			defer allocDir.Destroy()
+			executor := factory(testlog.HCLogger(t))
+			defer executor.Shutdown("", 0)
+
+			ps, cleanupHandle, err := executor.Launch(execCmd)
+			require.NoError(err)
+			require.NotZero(ps.Pid)
+
+			t.Logf("clean up handle is: %v", string(cleanupHandle))
+			t.Logf("process is %v", ps)
+
+			killed := make(chan *ProcessState, 1)
+			go func() {
+				defer close(killed)
+
+				proc, err := executor.Wait(context.Background())
+				if err == nil {
+					killed <- proc
+					return
+				}
+				t.Logf("original executor wait failed: %v", err)
+			}()
+
+			err = CleanupExecutor(testlog.HCLogger(t), cleanupHandle)
+			require.NoError(err)
+
+			select {
+			case proc, ok := <-killed:
+				require.True(ok, "original executor failed to wait")
+				require.Equal(int(syscall.SIGKILL), proc.Signal)
+			case <-time.After(4 * time.Second * time.Duration(tu.TestMultiplier())):
+				require.Fail("timed out waiting to be killed")
+			}
+		})
+	}
+}
+
+// TestExecutor_Cleanup_Dead asserts that CleanupExecutor detects
+// if task died
+func TestExecutor_Cleanup_Dead(pt *testing.T) {
+	pt.Parallel()
+	for name, factory := range executorFactories {
+		pt.Run(name, func(t *testing.T) {
+			require := require.New(t)
+			execCmd, allocDir := testExecutorCommand(t)
+			execCmd.Cmd = "/bin/sleep"
+			execCmd.Args = []string{"100"}
+			defer allocDir.Destroy()
+
+			executor := factory(testlog.HCLogger(t))
+			ps, cleanupHandle, err := executor.Launch(execCmd)
+			require.NoError(err)
+			require.NotZero(ps.Pid)
+
+			time.Sleep(2 * time.Second * time.Duration(tu.TestMultiplier()))
+			require.NoError(executor.Shutdown("SIGKILL", 100*time.Millisecond))
+			time.Sleep(2 * time.Second * time.Duration(tu.TestMultiplier()))
```

Comment:
The tests here fail without the sleeps or the grace period, presumably because of the race condition handled in https://github.com/hashicorp/nomad/pull/5429 , but I can dig further.